statesync/blocks.go

```diff
@@ -0,0 +1,282 @@
+	return q.doneCh
+}
+
+// VerifyNext pulls the next block off the pending queue and adds it to a
+// channel if it's already there or creates a waiter to add it to the
+// channel once it comes in. NOTE: This is assumed to
+// be a single thread as light blocks need to be sequentially verified.
+func (q *blockQueue) VerifyNext() <-chan lightBlockResponse {
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+	ch := make(chan lightBlockResponse, 1)
+
+	select {
+	case <-q.doneCh:
+		return ch
+	default:
+	}
+
+	if lb, ok := q.pending[q.verifyHeight]; ok {
+		ch <- lb
+		close(ch)
+		delete(q.pending, q.verifyHeight)
+	} else {
+		q.verifyCh = ch
+	}
+
+	return ch
+}
+
+// Retry is called when a dispatcher failed to fetch a light block or the
+// fetched light block failed verification. It signals to the queue to add the
+// height back to the request queue
+func (q *blockQueue) Retry(height int64) {
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+
+	select {
+	case <-q.doneCh:
+		return
+	default:
+	}
+
+	// we don't need to retry if this is below the terminal height
+	if q.terminal != nil && height < q.terminal.Height {
+		return
+	}
+
+	q.retries++
+	if q.retries >= q.maxRetries {
+		q.close()
+		return
+	}
+
+	if len(q.waiters) > 0 {
+		q.waiters[0] <- height
+		close(q.waiters[0])
+		q.waiters = q.waiters[1:]
+	} else {
+		q.failed = insertSorted(q.failed, height)
+	}
+}
+
+// Success is called when a light block has been successfully verified and
+// processed
+func (q *blockQueue) Success(height int64) {
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+	if q.hasFinalBlock() && q.verifyHeight == q.terminal.Height {
+		q.close()
+	}
+	q.verifyHeight--
+}
+
+// Close closes all channels associated with the queue
+func (q *blockQueue) Close() {
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+	q.close()
+}
+
+func (q *blockQueue) Error() error {
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+	if q.retries >= q.maxRetries {
+		return fmt.Errorf("failed to backfill blocks following reverse sync. Max retries exceeded (%d). "+
+			"Target height: %d, height reached: %d", q.maxRetries, q.stopHeight, q.verifyHeight)
+	}
+	return nil
+}
+
+func (q *blockQueue) close() {
+	close(q.doneCh)
+
+	// wait for the channel to be drained
+	select {
+	case <-q.doneCh:
+		return
+	default:
+	}
+
+	for _, ch := range q.waiters {
+		close(ch)
+	}
+	if q.verifyCh != nil {
+		close(q.verifyCh)
+	}
+}
+
+// nextFailed pops the next height in the list of failed heights
+func (q *blockQueue) nextFailed() int64 {
+	height := q.failed[0]
+	if len(q.failed) > 1 {
+		q.failed = q.failed[1:]
+	} else {
+		q.failed = []int64{}
+	}
+	return height
+}
+
+// hasFinalBlock check if the earliest block received has a height that is less than or equal to
+// the stop height and the last block time is before the stop time.
+func (q *blockQueue) hasFinalBlock() bool {
+	return q.terminal != nil
+}
+
+// insertAt inserts v into s at index i and returns the new slice.
+func insertAt(list []int64, i int, v int64) []int64 {
```

Comment:
Yeah that also makes sense

Reply:
Looking at this a second time, this is only used to keep track of the heights that failed. Putting the failed heights in order is a minor optimization and it could be an idea to just use a FIFO slice. I think I will leave this as is unless you have any strong doubts about it

Reply:
heaps are just ordered slices, at the end of the day, so I think it makes sense to use the library implementation, but yeah.